File: src/charmlint/linter.py
"""Core linter engine — loads charm context, discovers rules, runs them."""

import contextlib
import pathlib
from typing import Any

import yaml

from . import config as _config
from . import models
from . import rules as _rules


def _load_yaml(path: pathlib.Path) -> dict[str, Any]:
    """Load a YAML file, returning an empty dict on failure."""
    if not path.exists():
        return {}
    try:
        with path.open() as f:
            data = yaml.safe_load(f)
        return data if isinstance(data, dict) else {}
    except (yaml.YAMLError, OSError):
        return {}


def _collect_python_files(charm_dir: pathlib.Path) -> list[pathlib.Path]:
    """Collect all Python files in src/ and lib/ directories."""
    files: list[pathlib.Path] = []
    for subdir in ("src", "lib"):
        d = charm_dir / subdir
        if d.is_dir():
            files.extend(sorted(d.rglob("*.py")))
    return files


def _read_python_sources(python_files: list[pathlib.Path]) -> dict[pathlib.Path, str]:
    """Read all Python files into a content cache."""
    sources: dict[pathlib.Path, str] = {}
    for path in python_files:
        with contextlib.suppress(OSError):
            sources[path] = path.read_text(errors="replace")
    return sources


def _check_tests(charm_dir: pathlib.Path) -> tuple[bool, bool]:
    """Return (has_unit_tests, has_integration_tests)."""
    unit_dir = charm_dir / "tests" / "unit"
    integration_dir = charm_dir / "tests" / "integration"
    has_unit = unit_dir.is_dir() and bool(list(unit_dir.glob("test_*.py")))
    has_integration = integration_dir.is_dir() and bool(list(integration_dir.glob("test_*.py")))
    return has_unit, has_integration


def build_context(charm_dir: pathlib.Path) -> models.CharmContext:
    """Load all charm data into a CharmContext for rule evaluation."""
    charm_dir = charm_dir.resolve()

    # Load metadata from charmcraft.yaml or legacy metadata.yaml.
    metadata = _load_yaml(charm_dir / "charmcraft.yaml")
    if not metadata:
        metadata = _load_yaml(charm_dir / "metadata.yaml")

    # Load actions (charmcraft.yaml or actions.yaml).
    actions: dict[str, Any] = metadata.get("actions", {})
    if not actions:
        actions_data = _load_yaml(charm_dir / "actions.yaml")
        actions = actions_data if isinstance(actions_data, dict) else {}

    # Load config options (charmcraft.yaml or config.yaml).
    config_section = metadata.get("config", {})
    if isinstance(config_section, dict) and config_section.get("options"):
        config_options = config_section["options"]
    elif isinstance(config_section, dict) and config_section:
        config_options = config_section
    else:
        config_data = _load_yaml(charm_dir / "config.yaml")
        config_options = config_data.get("options", config_data) if config_data else {}

    # Collect Python files and read their contents.
    python_files = _collect_python_files(charm_dir)
    python_sources = _read_python_sources(python_files)

    # Read README.
    readme_content = ""
    readme_path = charm_dir / "README.md"
    if readme_path.exists():
        with contextlib.suppress(OSError):
            readme_content = readme_path.read_text(errors="replace")

    has_unit, has_integration = _check_tests(charm_dir)

    return models.CharmContext(
        charm_dir=charm_dir,
        metadata=metadata,
        actions=actions,
        config_options=config_options,
        python_files=python_files,
        python_sources=python_sources,
        readme_content=readme_content,
        has_tests_unit=has_unit,
        has_tests_integration=has_integration,
    )


def _should_run_rule(rule: _rules.Rule, config: _config.LintConfig) -> bool:
    """Determine whether a rule should run given the config."""
    rule_id = rule.id
    category = rule_id.rstrip("0123456789")

    # Explicit disable via severity override.
    if config.severity_overrides.get(rule_id) == "off":
        return False

    # Category-level disable.
    if config.severity_overrides.get(category) == "off":
        return False

    # If select is set, only run rules in those categories.
    if config.select and category not in config.select:
        return False

    # If ignore contains this specific rule or category, skip it.
    return not (rule_id in config.ignore or category in config.ignore)


def _effective_severity(rule: _rules.Rule, config: _config.LintConfig) -> models.Severity | None:
    """Resolve the effective severity for a rule, applying config overrides."""
    rule_id = rule.id
    override = config.severity_overrides.get(rule_id)
    if override and override != "off":
        try:
            return models.Severity(override)
        except ValueError:
            pass
    return None


def lint(
    charm_dir: pathlib.Path,
    config: _config.LintConfig | None = None,
) -> models.LintReport:
    """Run all enabled rules against a charm directory.

    This is the main public API.
    """
    if config is None:
        config = _config.LintConfig()

    # Ensure all rule modules are imported so rules register themselves.
    _ensure_rules_loaded()

    context = build_context(charm_dir)

    if not context.metadata:
        return models.LintReport(
            charm_dir=charm_dir,
            diagnostics=[
                models.Diagnostic(
                    rule_id="FATAL",
                    severity=models.Severity.ERROR,
                    message=(
                        "No charmcraft.yaml or metadata.yaml found — is this a charm directory?"
                    ),
                )
            ],
        )

    all_diagnostics: list[models.Diagnostic] = []

    for rule in _rules.get_all_rules().values():
        if not _should_run_rule(rule, config):
            continue

        diagnostics = rule.check(context)

        # Apply severity overrides.
        override = _effective_severity(rule, config)
        if override is not None:
            diagnostics = [
                models.Diagnostic(
                    rule_id=d.rule_id,
                    severity=override,
                    message=d.message,
                    path=d.path,
                    line=d.line,
                    fix_hint=d.fix_hint,
                )
                for d in diagnostics
            ]

        # Filter by minimum severity.
        if config.min_severity:
            severity_order = {
                models.Severity.ERROR: 0,
                models.Severity.WARNING: 1,
                models.Severity.INFO: 2,
            }
            min_order = severity_order.get(config.min_severity, 2)
            diagnostics = [
                d for d in diagnostics if severity_order.get(d.severity, 2) <= min_order
            ]

        all_diagnostics.extend(diagnostics)

    return models.LintReport(charm_dir=charm_dir, diagnostics=all_diagnostics)


_rules_loaded = False


def _ensure_rules_loaded() -> None:
    """Import all rule modules so their Rule subclasses register."""
    global _rules_loaded  # noqa: PLW0603
    if _rules_loaded:
        return
    _rules_loaded = True

    from .rules import (
        actions,  # noqa: F401
        charmcraft_compat,  # noqa: F401
        config_quality,  # noqa: F401
        deprecated,  # noqa: F401
        documentation,  # noqa: F401
        libraries,  # noqa: F401
        metadata,  # noqa: F401
        observability,  # noqa: F401
        security,  # noqa: F401
        status,  # noqa: F401
        structure,  # noqa: F401
        testing,  # noqa: F401
    )

File: src/charmlint-rs/src/linter.rs
//! Core linter engine — loads context, runs rules, applies filters.

use crate::config::LintConfig;
use crate::context;
use crate::models::{Diagnostic, LintReport, Severity};
use crate::rules;
use std::path::Path;

/// Extract category from rule ID (strip trailing digits).
fn rule_category(rule_id: &str) -> String {
    rule_id.trim_end_matches(|c: char| c.is_ascii_digit()).to_string()
}

/// Check whether a rule should run given the config.
fn should_run_rule(rule_id: &str, config: &LintConfig) -> bool {
    let category = rule_category(rule_id);

    // Explicit disable via severity override.
    if config.severity_overrides.get(rule_id).map(|s| s.as_str()) == Some("off") {
        return false;
    }
    if config
        .severity_overrides
        .get(&category)
        .map(|s| s.as_str())
        == Some("off")
    {
        return false;
    }

    // If select is set, only run rules in those categories.
    if !config.select.is_empty() && !config.select.contains(&category) {
        return false;
    }

    // If ignore contains this specific rule or category, skip it.
    !config.ignore.contains(&rule_id.to_string()) && !config.ignore.contains(&category)
}

/// Resolve the effective severity for a rule, applying config overrides.
fn effective_severity(rule_id: &str, config: &LintConfig) -> Option<Severity> {
    if let Some(override_val) = config.severity_overrides.get(rule_id) {
        if override_val != "off" {
            return Severity::from_str_loose(override_val);
        }
    }
    None
}

/// Run all enabled rules against a charm directory.
pub fn lint(charm_dir: &Path, config: &LintConfig) -> LintReport {
    let ctx = context::build_context(charm_dir);

    if ctx.metadata.is_empty() {
        return LintReport::new(
            charm_dir,
            vec![Diagnostic {
                rule_id: "FATAL".to_string(),
                severity: Severity::Error,
                message:
                    "No charmcraft.yaml or metadata.yaml found — is this a charm directory?"
                        .to_string(),
                path: None,
                line: None,
                fix_hint: None,
            }],
        );
    }

    let all_diagnostics = rules::run_all(&ctx);

    // Filter and transform diagnostics.
    let mut filtered = Vec::new();
    for d in all_diagnostics {
        if !should_run_rule(&d.rule_id, config) {
            continue;
        }

        // Apply severity overrides.
        let d = if let Some(sev) = effective_severity(&d.rule_id, config) {
            Diagnostic {
                severity: sev,
                ..d
            }
        } else {
            d
        };

        // Filter by minimum severity.
        if let Some(min_sev) = &config.min_severity {
            if d.severity.order() > min_sev.order() {
                continue;
            }
        }

        filtered.push(d);
    }

    LintReport::new(charm_dir, filtered)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn write(path: &Path, contents: &str) {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent).unwrap();
        }
        std::fs::write(path, contents).unwrap();
    }

    fn minimal_charm() -> tempfile::TempDir {
        let dir = tempfile::tempdir().unwrap();
        std::fs::create_dir(dir.path().join("src")).unwrap();
        write(&dir.path().join("charmcraft.yaml"), "name: test\n");
        dir
    }

    #[test]
    fn rule_category_strips_trailing_digits() {
        assert_eq!(rule_category("COS005"), "COS");
        assert_eq!(rule_category("META001"), "META");
        assert_eq!(rule_category("STR"), "STR");
    }

    #[test]
    fn should_run_rule_respects_off_override_for_specific_id() {
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS005".to_string(), "off".to_string());
        assert!(!should_run_rule("COS005", &config));
        assert!(should_run_rule("COS001", &config));
    }

    #[test]
    fn should_run_rule_respects_off_override_for_category() {
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS".to_string(), "off".to_string());
        assert!(!should_run_rule("COS005", &config));
        assert!(!should_run_rule("COS001", &config));
    }

    #[test]
    fn should_run_rule_with_select_only_runs_listed_categories() {
        let config = LintConfig {
            select: vec!["META".to_string()],
            ..LintConfig::default()
        };
        assert!(should_run_rule("META001", &config));
        assert!(!should_run_rule("COS005", &config));
    }

    #[test]
    fn should_run_rule_respects_ignore_list() {
        let config = LintConfig {
            ignore: vec!["STR003".to_string()],
            ..LintConfig::default()
        };
        assert!(!should_run_rule("STR003", &config));
        assert!(should_run_rule("STR001", &config));
    }

    #[test]
    fn effective_severity_resolves_from_override() {
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS001".to_string(), "error".to_string());
        assert_eq!(effective_severity("COS001", &config), Some(Severity::Error));
    }

    #[test]
    fn effective_severity_ignores_off() {
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS001".to_string(), "off".to_string());
        assert!(effective_severity("COS001", &config).is_none());
    }

    #[test]
    fn lint_returns_fatal_when_no_metadata_found() {
        let dir = tempfile::tempdir().unwrap();
        let report = lint(dir.path(), &LintConfig::default());
        assert_eq!(report.errors, 1);
        assert_eq!(report.diagnostics[0].rule_id, "FATAL");
    }

    #[test]
    fn lint_with_select_only_emits_selected_categories() {
        let dir = minimal_charm();
        let config = LintConfig {
            select: vec!["META".to_string()],
            ..LintConfig::default()
        };
        let report = lint(dir.path(), &config);
        for d in &report.diagnostics {
            assert!(d.rule_id.starts_with("META"), "unexpected rule {}", d.rule_id);
        }
    }

    #[test]
    fn lint_with_ignore_excludes_named_rule() {
        let dir = minimal_charm();
        let config = LintConfig {
            ignore: vec!["TEST001".to_string()],
            ..LintConfig::default()
        };
        let report = lint(dir.path(), &config);
        assert!(!report.diagnostics.iter().any(|d| d.rule_id == "TEST001"));
    }

    #[test]
    fn lint_severity_override_promotes_rule_to_error() {
        let dir = minimal_charm();
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS001".to_string(), "error".to_string());
        let report = lint(dir.path(), &config);
        let cos001: Vec<&Diagnostic> =
            report.diagnostics.iter().filter(|d| d.rule_id == "COS001").collect();
        assert!(!cos001.is_empty());
        assert_eq!(cos001[0].severity, Severity::Error);
    }

    #[test]
    fn lint_disable_via_off_removes_rule() {
        let dir = minimal_charm();
        let mut config = LintConfig::default();
        config
            .severity_overrides
            .insert("COS001".to_string(), "off".to_string());
        let report = lint(dir.path(), &config);
        assert!(!report.diagnostics.iter().any(|d| d.rule_id == "COS001"));
    }

    #[test]
    fn lint_min_severity_filters_lower_severities() {
        let dir = minimal_charm();
        let config = LintConfig {
            min_severity: Some(Severity::Error),
            ..LintConfig::default()
        };
        let report = lint(dir.path(), &config);
        for d in &report.diagnostics {
            assert_eq!(d.severity, Severity::Error, "rule {}", d.rule_id);
        }
    }
}
